client/src/controllers/SwapController.ts

```diff
@@ -0,0 +1,288 @@
+      throw new Error(errors.join(', '));
+    }
+
+    return targetElement as HTMLElement;
+  }
+
+  /**
+   * Toggle the visual spinner icon if available and ensure content about
+   * to be replaced is flagged as busy.
+   */
+  loadingValueChanged(isLoading: boolean) {
+    if (isLoading) {
+      this.targetElement?.setAttribute('aria-busy', 'true');
+      this.iconElement?.setAttribute('href', '#icon-spinner');
+    } else {
+      this.targetElement?.removeAttribute('aria-busy');
+      this.iconElement?.setAttribute('href', this.iconValue);
+    }
+  }
+
+  /**
+   * Abort any existing requests & set up new abort controller, then fetch and patch
+   * in the HTML results, dispatching an event and handling any clean up.
+   * Cancel any in progress results request using the AbortController so that
+   * a faster response does not replace an in flight request.
+   */
+  async request(param: string | CustomEvent<{ url: string }>) {
+    if (this.abortController) this.abortController.abort();
+    this.abortController = new AbortController();
+    const { signal } = this.abortController;
+
+    /** Parse a request URL from the supplied param, as a string or inside a custom event */
+    const requestUrl =
+      typeof param === 'string' ? param : param.detail.url || '';
+
+    this.loadingValue = true;
+
+    const beginEvent = this.dispatch('begin', {
+      cancelable: true,
+      detail: { requestUrl },
+      // Stimulus dispatch target element type issue https://github.com/hotwired/stimulus/issues/642
+      target: this.targetElement as HTMLInputElement,
+    }) as CustomEvent<{ requestUrl: string }>;
+
+    if (beginEvent.defaultPrevented) return Promise.resolve();
+
+    return fetch(requestUrl, {
+      headers: { 'x-requested-with': 'XMLHttpRequest' },
+      signal,
+    })
+      .then((response) => {
+        if (!response.ok) {
+          throw new Error(`HTTP error! Status: ${response.status}`);
+        }
+        return response.text();
+      })
+      .then((results) => {
+        const targetElement = this.targetElement as HTMLElement;
+        targetElement.innerHTML = results;
+        this.dispatch('success', {
+          cancelable: false,
+          detail: { requestUrl, results },
+          // Stimulus dispatch target element type issue https://github.com/hotwired/stimulus/issues/642
+          target: targetElement as HTMLInputElement,
+        });
+        return results;
+      })
+      .catch((error) => {
+        if (signal.aborted) return;
+        this.dispatch('error', {
+          cancelable: false,
+          detail: { error, requestUrl },
+          // Stimulus dispatch target element type issue https://github.com/hotwired/stimulus/issues/642
+          target: this.targetElement as HTMLInputElement,
+        });
+        // eslint-disable-next-line no-console
+        console.error(`Error fetching ${requestUrl}`, error);
+      })
+      .finally(() => {
+        this.loadingValue = false;
```

Comment:
I think we should ensure that the promise is for the latest request, as otherwise this would set the `loadingValue` to `false` when there is a request that is aborted after a new one has started and is in progress.

I was able to reproduce this issue by throttling my network to be really slow via devtools (e.g. 1 kBps, or maybe 5kBps is enough): there is an ongoing request, but because a previous request is aborted, it sets the `loadingValue` to `false`.

The following seems to fix this:

```suggestion
        if (signal === this.abortController?.signal) {
          this.loadingValue = false;
        }
```